spec/services/google/search_service_spec.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'support/http_client_helpers'
+
+RSpec.describe Google::SearchService, type: :service do
+  describe '#search!' do
+    context 'when search a keyword' do
+      it 'perform a request against www.google.com' do
+        outer_env = nil
+        HttpClientHelpers.stub_http_adapter do |adapter, connection|
+          allow(described_class).to receive(:http_client).and_return(connection)
+          adapter.get('/search') do |env|
+            outer_env = env
+            200
+          end
+        end
+
+        described_class.search!('hello world')
+        expect(outer_env.url.host).to include('www.google.com')
```

Comment:
Many of the tests in this file test implementation details, which is a code smell. Instead, only test the behavior (the B in BDD testing). The tests should not know too much about the implementation but only care about the output of the code. It makes tests easier to maintain as we can swap the implementation while keeping the same behavior without breaking the tests. 

So to illustrate this point, there is no value in checking which host the code requests to. What matters is what the request returns. So here for simplicity we can also use `vcr` wand not use any stubs.

Reply:
I agree with your opinion regarding the target host test case

To make sure we're on a same page. This test case
```ruby
it 'raise GoogleScraperRuby::Errors::SearchServiceError if the request returns a 4xx status code' do
  HttpClientHelpers.stub_http_adapter do |adapter, connection|
    allow(described_class).to receive(:http_client).and_return(connection)
    adapter.get('/search') { [429, {}, 'Too Many Requests'] }
  end

  expect do
    described_class.search!('hello world')
  end.to raise_error(GoogleScraperRuby::Errors::SearchServiceError)
end
```

also knows a lot about SearchService implementation (the face that it calls `http_client` to build the http client). So a better solution is to use a vcr which returns a 429 response. This will make sure the test is still correct even if we change the HTTP client library. Is it correct?

Reply:
also, I'm struggling to decide if this is a behavior we want to test:

```ruby
it "raise GoogleScraperRuby::Errors::SearchServiceError if it can't connect to the host" do
  HttpClientHelpers.stub_http_adapter do |adapter, connection|
    allow(described_class).to receive(:http_client).and_return(connection)
    adapter.get('/search') { raise Faraday::ConnectionFailed }
  end

  expect do
    described_class.search!('hello world')
  end.to raise_error(GoogleScraperRuby::Errors::SearchServiceError)
end
```

If it's worth testing then how should we do it. The current implementation is leaking implementation details i.e.  the `Faraday::ConnectionFailed` error